### PR TITLE
[soundcloud] Add Soundcloud search extractor

### PR DESCRIPTION
**Pull request by reiv**
It would be nice to be able to search for tracks on Soundcloud directly from youtube-dl.

I've added a `SoundcloudSearchIE` extractor that is invoked with `scsearch:"query"`. It uses the Soundcloud API v2.

Thanks.

### CODE REVIEW

File: youtube_dl/extractor/soundcloud.py [hunk 1]
```diff
@@ -469,3 +472,86 @@ def _real_extract(self, url):
...
+                break
+
+            next_url = response.get(u'next_href', None)
+
+    def _get_n_results(self, query, n):
+
+        results = []
+
+        tracks = self._get_collection('/search/tracks',
+            collection_id='Query "{}"'.format(query),
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: `{}` doesn't work in python 2.6.

File: youtube_dl/extractor/soundcloud.py [hunk 2]
```diff
@@ -469,3 +472,86 @@ def _real_extract(self, url):
...
+            q=query.encode('utf-8'))
+
+        for _ in range(n):
+            try:
+                track = next(tracks)
+            except StopIteration:
+                break
+            uri = track[u'uri']
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: Don't use `u`.

File: youtube_dl/extractor/soundcloud.py [hunk 3]
```diff
@@ -469,3 +472,86 @@ def _real_extract(self, url):
...
+            title = track[u'title']
+            username = track[u'user'][u'username']
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: It shouldn't fail if `user` or `username` is missing.

File: youtube_dl/extractor/soundcloud.py [hunk 4]
```diff
@@ -469,3 +472,86 @@ def _real_extract(self, url):
...
+    _RESULTS_PER_PAGE = 50
+
+    def _get_collection(self, endpoint, collection_id, **query):
+        import itertools
+
+        query['limit'] = self._RESULTS_PER_PAGE
+        query['client_id'] = self._CLIENT_ID
+        query['linked_partitioning'] = '1'
+
+        api_base_url = '{0}//api-v2.soundcloud.com'.format(self.http_scheme())
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: This makes no sense since it redirected to https anyway.

File: youtube_dl/extractor/soundcloud.py [hunk 5]
```diff
@@ -469,3 +472,86 @@ def _real_extract(self, url):
...
+            'title': 'post-avant jazzcore',
+        },
+        'playlist_count': 15,
+    }]
+
+    _SEARCH_KEY = 'scsearch'
+    _RESULTS_PER_PAGE = 50
+
+    def _get_collection(self, endpoint, collection_id, **query):
+        import itertools
```
Comment (jaimeMF) [inline youtube_dl/extractor/soundcloud.py]: This is already imported and (in general) you should only use `import`s at the top level.

File: youtube_dl/extractor/soundcloud.py [hunk 6]
```diff
@@ -469,3 +472,86 @@ def _real_extract(self, url):
...
+
+    def _get_n_results(self, query, n):
+
+        results = []
+
+        tracks = self._get_collection('/search/tracks',
+            collection_id='Query "{}"'.format(query),
+            q=query.encode('utf-8'))
+
+        for _ in range(n):
```
Comment (jaimeMF) [inline youtube_dl/extractor/soundcloud.py]: Is there some reason why you can't just use `for track in tracks:` here?

Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: `_get_collection()` is a generator that loads lazily from a paginated source. Writing `for track in tracks:` would load all results, but we want to stop at `n`. This way we avoid making unnecessary requests.

File: youtube_dl/extractor/soundcloud.py (hunk 5 above)
Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: Fixed.

File: youtube_dl/extractor/soundcloud.py (hunk 3 above)
Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: Turns out it wasn't necessary anyway.

File: youtube_dl/extractor/soundcloud.py (hunk 4 above)
Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: @dstftw: Okay, removed.

File: youtube_dl/extractor/soundcloud.py (hunk 2 above)
Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: Fixed.

File: youtube_dl/extractor/soundcloud.py (hunk 6 above)
Comment (jaimeMF) [inline youtube_dl/extractor/soundcloud.py]: I missed that, then you should either use [itertools.islice](https://docs.python.org/2/library/itertools.html#itertools.islice) or
```python
for i, track in enumerate(tracks):
    if i >= n:
        break
```

Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: Thanks, I rewrote it with `itertools.islice`.

File: youtube_dl/extractor/soundcloud.py [hunk 7]
```diff
@@ -469,3 +472,70 @@ def _real_extract(self, url):
...
+        tracks = self._get_collection('/search/tracks',
+            collection_id='Query "{0}"'.format(query),
+            q=query.encode('utf-8'))
+
+        results = [self.url_result(url=track['uri'])
+            for track in itertools.islice(tracks, n)]
+
+        if not results:
+            raise ExtractorError(
+                '[soundcloud] No track results', expected=True)
```
Comment (yan12125) [inline youtube_dl/extractor/soundcloud.py]: Usually we use syntaxes like 'soundcloud said: ' but not brackets in error messages.

Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: Changed.

File: youtube_dl/extractor/soundcloud.py [hunk 8]
```diff
@@ -469,3 +472,70 @@ def _real_extract(self, url):
             'description': data.get('description'),
             'entries': entries,
         }
+
+
+class SoundcloudSearchIE(SearchInfoExtractor, SoundcloudIE):
+    IE_NAME = 'soundcloud:search'
+    IE_DESC = 'Soundcloud search'
+    _MAX_RESULTS = 200
```
Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: Any particular reason for limiting this at all?

Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: @dstftw: it's in the API docs. 200 is the maximum number of items returned in a collection. See https://developers.soundcloud.com/docs#pagination

Comment (dstftw) [inline youtube_dl/extractor/soundcloud.py]: >The number of items in the collection returned is limited to 50 by default with a maximum value of 200

Collection is a single page thus 200 is the limit for a single page. So, `_RESULTS_PER_PAGE` should be set to 200 and `_MAX_RESULTS` to `float('inf')`.

Comment (reiv) [inline youtube_dl/extractor/soundcloud.py]: I took `limit` to mean the limit on the number of results altogether, but your interpretation makes more sense.

**Comment by dstftw** [comment]
Cherry-picked. Thanks.

**Comment by dstftw** [comment]
Do you want to be credited as one of the authors of youtube-dl? If so under what name?

**Comment by reiv** [comment]
Cheers! You can use my github username.

**Pull request closed without merging**